pkg/component/gardenerapiserver/gardener_apiserver.go

```diff
@@ -0,0 +1,205 @@
+type Values struct {
+	apiserver.Values
+	// Image is the container images used for the gardener-apiserver pods.
+	Image string
+}
+
+// New creates a new instance of DeployWaiter for the gardener-apiserver.
+func New(client client.Client, namespace string, secretsManager secretsmanager.Interface, values Values) Interface {
+	return &gardenerAPIServer{
+		client:         client,
+		namespace:      namespace,
+		secretsManager: secretsManager,
+		values:         values,
+	}
+}
+
+type gardenerAPIServer struct {
+	client         client.Client
+	namespace      string
+	secretsManager secretsmanager.Interface
+	values         Values
+}
+
+func (g *gardenerAPIServer) Deploy(ctx context.Context) error {
+	var (
+		runtimeRegistry = managedresources.NewRegistry(operatorclient.RuntimeScheme, operatorclient.RuntimeCodec, operatorclient.RuntimeSerializer)
+	)
+
+	runtimeResources, err := runtimeRegistry.AddAllAndSerialize()
+	if err != nil {
+		return err
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.CreateForSeed(ctx, g.client, g.namespace, managedResourceNameRuntime, false, runtimeResources); err != nil {
+		return err
+	}
+	if err := managedresources.WaitUntilHealthy(timeoutCtx, g.client, g.namespace, managedResourceNameRuntime); err != nil {
+		return err
+	}
+
+	var (
+		virtualRegistry = managedresources.NewRegistry(operatorclient.VirtualScheme, operatorclient.VirtualCodec, operatorclient.VirtualSerializer)
+	)
+
+	virtualResources, err := virtualRegistry.AddAllAndSerialize()
+	if err != nil {
+		return err
+	}
+
+	timeoutCtx, cancel = context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.CreateForShoot(ctx, g.client, g.namespace, managedResourceNameVirtual, managedresources.LabelValueGardener, false, virtualResources); err != nil {
+		return err
+	}
+	return managedresources.WaitUntilHealthy(timeoutCtx, g.client, g.namespace, managedResourceNameVirtual)
+}
+
+func (g *gardenerAPIServer) Destroy(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.DeleteForShoot(ctx, g.client, g.namespace, managedResourceNameVirtual); err != nil {
+		return err
+	}
+	if err := managedresources.WaitUntilDeleted(timeoutCtx, g.client, g.namespace, managedResourceNameVirtual); err != nil {
+		return err
+	}
+
+	timeoutCtx, cancel = context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.DeleteForSeed(ctx, g.client, g.namespace, managedResourceNameRuntime); err != nil {
+		return err
+	}
+	return managedresources.WaitUntilDeleted(timeoutCtx, g.client, g.namespace, managedResourceNameRuntime)
+}
+
+func (g *gardenerAPIServer) Wait(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
```

Comment:
Should `timeoutCtx` not be created after the `if` block? It seems like we wait sequentially for the runtime and virtual managed resources. However, unlike all other operations here the timeout counts for both operations together, but not separately for each individually.